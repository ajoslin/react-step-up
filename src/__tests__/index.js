const React = require('react')
const {mount} = require('enzyme')
const Stepper = require('../')

function setup ({render = () => <div />, ...props} = {}) {
  let renderArg
  const renderSpy = jest.fn(arg => {
    renderArg = arg
    return render(arg)
  })
  const wrapper = mount(<Stepper {...props} render={renderSpy} />)
  return {renderSpy, wrapper, ...renderArg}
}

test('value defaults to 0', () => {
  const {value} = setup()
  expect(value).toBe(0)
})

test('value follows defaultValue', () => {
  const {value} = setup({defaultValue: 33})
  expect(value).toBe(33)
})

test('getFormProps', () => {
  const {getFormProps} = setup({})
  expect(getFormProps()).toMatchSnapshot()
})

test('getInputProps', () => {
  const {wrapper, getInputProps} = setup({})
  expect(getInputProps().value).toBe(0)
  expect(getInputProps()).toMatchSnapshot()
  wrapper.setState({ focused: true })
  expect(getInputProps().value).toBe(undefined)
  expect(getInputProps()).toMatchSnapshot()
})

test('getIncrementProps', () => {
  const {getIncrementProps} = setup()
  expect(getIncrementProps()).toMatchSnapshot()
})

test('getDecrementProps', () => {
  const {getDecrementProps} = setup()
  expect(getDecrementProps()).toMatchSnapshot()
})

test('setValue is between [min, max]', () => {
  const {wrapper, setValue} = setup({min: 0, max: 1})

  setValue(2)
  expect(wrapper.state('value')).toBe(1)
  setValue(-1)
  expect(wrapper.state('value')).toBe(0)
  setValue(1)
  expect(wrapper.state('value')).toBe(1)
  setValue(0)
  expect(wrapper.state('value')).toBe(0)
})

test('increment/decrement are capped at min/max', () => {
  const {wrapper, increment, decrement} = setup({min: 0, max: 1})

  expect(wrapper.state('value')).toBe(0)
  increment()
  expect(wrapper.state('value')).toBe(1)
  increment()
  expect(wrapper.state('value')).toBe(1)
  decrement()
  expect(wrapper.state('value')).toBe(0)
  decrement()
  expect(wrapper.state('value')).toBe(0)
})

describe('enableReinitialize', () => {
  test('true: value is updated to new default if defaultValue changes and value has not been modified', () => {
    const {wrapper} = setup({defaultValue: 33, enableReinitialize: true})

    expect(wrapper.state('value')).toBe(33)
    wrapper.setProps({defaultValue: 42})
    expect(wrapper.state('value')).toBe(42)
  })

  test('true: value is not updated to new default if defaultValue changes and value has been modified', () => {
    const {wrapper} = setup({defaultValue: 33, enableReinitialize: true})

    expect(wrapper.state('value')).toBe(33)
    wrapper.setState({value: 418})
    wrapper.setProps({defaultValue: 42})
    expect(wrapper.state('value')).toBe(418)
  })

  test('false: value remains unchanged if defaultValue changes', () => {
    const {wrapper} = setup({defaultValue: 33})

    expect(wrapper.state('value')).toBe(33)
    wrapper.setProps({defaultValue: 42})
    expect(wrapper.state('value')).toBe(33)
  })
})
