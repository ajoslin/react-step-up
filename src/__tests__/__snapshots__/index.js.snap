// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDecrementProps 1`] = `
Object {
  "onClick": [Function],
}
`;

exports[`getFormProps 1`] = `
Object {
  "onSubmit": [Function],
}
`;

exports[`getIncrementProps 1`] = `
Object {
  "onClick": [Function],
}
`;

exports[`getInputProps 1`] = `
Object {
  "onBlur": [Function],
  "onFocus": [Function],
  "pattern": "[0-9]*",
  "ref": [Function],
  "type": "text",
  "value": 0,
}
`;

exports[`getInputProps 2`] = `
Object {
  "onBlur": [Function],
  "onFocus": [Function],
  "pattern": "[0-9]*",
  "ref": [Function],
  "type": "text",
}
`;
